optuna/samplers/_grid.py

```diff
@@ -98,6 +100,8 @@ def __init__(self, search_space: Mapping[str, Sequence[GridValueType]]) -> None:
 
         self._search_space = collections.OrderedDict()
         for param_name, param_values in sorted(search_space.items(), key=lambda x: x[0]):
+            param_values = cast(SortableParamValueSequenceType, param_values)
```

Comment:
I'm not sure if we have to check types of values in param_values explicitly.
I introduce `SortableParamValueSequenceType` without explicitly checking now.

Reply:
So we need it for https://github.com/optuna/optuna/blob/master/optuna/samplers/_grid.py#L221-L223. We can probably replace it with a set comparison, do you want to do that in this PR? It might stretch the scope of this PR somewhat since we currently seemingly allow duplicate items in the grid. It also affect performance, although this in practice shouldn't be an issue.

Reply:
Thank you for the comment and sorry for the late reply. As you pointed out, I also think we can implement `_same_search_space` with a set comparison. Let me make another PR, as I think it would be better to focus only on mypy fixing in this PR.